strawberryfields/compilers/xcov.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2019-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""General state compiler for the X class of circuits."""
+
+import copy
+
+import numpy as np
+from thewalrus.quantum import Amat
+from thewalrus.symplectic import expand
+
+import strawberryfields as sf
+from strawberryfields.decompositions import takagi
+from strawberryfields.program_utils import CircuitError, Command
+import strawberryfields.ops as ops
+
+from .compiler import Compiler
+from .gbs import GBS
+from .gaussian_unitary import GaussianUnitary
+
+
+class Xcov(Compiler):
+    """General state compiler for the X class of circuits.
+
+    An important property of this compilation routine is that it is done at the covariance matrix level.
+    This implies that one should not use it to compare the interferometers of a given circuit since they may
+    differ by permutations in the unitary and the squeezing parameters.
+    """
+
+    short_name = "Xcov"
+    interactive = False
+
+    primitives = {
+        "S2gate",
+        "Sgate",
+        "MeasureFock",
+        "Rgate",
+        "BSgate",
+        "MZgate",
+        "Interferometer",
+    }
+
+    decompositions = {
+        "BipartiteGraphEmbed": {"mesh": "rectangular_symmetric", "drop_identity": False},
+    }
+
+    def compile(self, seq, registers):
+        # the number of modes in the provided program
+        n_modes = len(registers)
+        half_n_modes = n_modes // 2
+
+        # Number of modes must be even
+        if n_modes % 2 != 0:
+            raise CircuitError("The X series only supports programs with an even number of modes.")
+
+        # Call the GBS compiler to do basic measurement validation.
+        # The GBS compiler also merges multiple measurement commands
+        # into a single MeasureFock command at the end of the circuit.
+        seq = GBS().compile(seq, registers)
+
+        # ensure that all modes are measured
+        if len(seq[-1].reg) != n_modes:
+            raise CircuitError("All modes must be measured.")
+
+        meas_seq = [seq[-1]]
+
+        if len(seq) == 1:
+            # program consists only of measurements
+            sq_seq = [
+                Command(ops.S2gate(0), [registers[i], registers[i + half_n_modes]])
+                for i in range(half_n_modes)
+            ]
```

Comment:
how come there's squeezing for when only measurements were defined?

Reply:
hey Antal, I am not sure I understand your question.

Reply:
Hi Nico, this is a branch where the user only added measurements, hence the program only consists of measurements. I was unsure why commands for `S2gate` are added here, but I would assume it is just because of how the circuit should look like. One thing I missed to parse previously was the squeezing amount being 0.

Reply:
Understood. Yeah it is to match the chip description and make sure that a correctly formatted blackbird script is returned.